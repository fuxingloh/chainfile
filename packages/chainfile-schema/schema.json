{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "allOf": [
    {
      "$ref": "#/definitions/Chainfile"
    }
  ],
  "definitions": {
    "Chainfile": {
      "type": "object",
      "properties": {
        "$schema": {
          "type": "string"
        },
        "description": {
          "type": "string",
          "minLength": 1,
          "maxLength": 1024
        },
        "caip2": {
          "type": "string",
          "pattern": "^[-a-z0-9]{3,8}:[-_a-zA-Z0-9]{1,32}$",
          "description": "CAIP-2 Chain ID of the blockchain network this chainfile is for."
        },
        "name": {
          "type": "string",
          "description": "Describes this chainfile.",
          "minLength": 1,
          "maxLength": 128
        },
        "params": {
          "type": "object",
          "maxProperties": 30,
          "patternProperties": {
            "^[a-z]+(_[a-z0-9]+)*$": {
              "oneOf": [
                {
                  "type": "string"
                },
                {
                  "$ref": "#/definitions/ParamOptions"
                }
              ]
            }
          },
          "additionalProperties": false
        },
        "volumes": {
          "type": "object",
          "patternProperties": {
            "^(?!chainfile)[a-z0-9][a-z0-9-]{0,28}[a-z0-9]$": {
              "$ref": "#/definitions/Volume"
            }
          },
          "additionalProperties": false
        },
        "containers": {
          "type": "object",
          "minProperties": 1,
          "maxProperties": 6,
          "patternProperties": {
            "^(?!chainfile)[a-z0-9][a-z0-9-]{0,28}[a-z0-9]$": {
              "$ref": "#/definitions/Container"
            }
          },
          "additionalProperties": false
        }
      },
      "required": ["caip2", "name", "containers"],
      "additionalProperties": false
    },
    "ParamOptions": {
      "type": "object",
      "properties": {
        "secret": {
          "type": "boolean",
          "description": "Mark the value as a secret"
        },
        "default": {
          "anyOf": [
            {
              "type": "string"
            },
            {
              "type": "object",
              "properties": {
                "random": {
                  "type": "object",
                  "description": "Generate a random value for orchestration that supports it.",
                  "properties": {
                    "bytes": {
                      "type": "number",
                      "minimum": 1,
                      "maximum": 256,
                      "multipleOf": 1
                    },
                    "encoding": {
                      "enum": ["hex", "base64", "base64url"]
                    }
                  },
                  "required": ["bytes", "encoding"],
                  "additionalProperties": false
                }
              },
              "additionalProperties": false
            }
          ]
        },
        "description": {
          "type": "string",
          "minLength": 1,
          "maxLength": 1024
        }
      },
      "additionalProperties": false
    },
    "ParamReference": {
      "type": "object",
      "properties": {
        "$param": {
          "type": "string",
          "description": "Name of the value to reference in the params section.",
          "pattern": "^[a-z]+(_[a-z0-9]+)*$"
        }
      },
      "required": ["$param"],
      "additionalProperties": false
    },
    "Volume": {
      "type": "object",
      "properties": {
        "type": {
          "enum": ["persistent", "ephemeral"]
        },
        "size": {
          "type": "string",
          "pattern": "^[1-9][0-9]*[MGT]i$"
        },
        "expansion": {
          "type": "object",
          "description": "Optional growth rate for volumes for evolving storage needs. [(size) + (rate * time)]",
          "properties": {
            "startFrom": {
              "type": "string",
              "pattern": "^[0-9]{4}-[0-9]{2}-[0-9]{2}$",
              "description": "Formatted as YYYY-MM-DD to denote the date the growth rate starts."
            },
            "monthlyRate": {
              "type": "string",
              "pattern": "^[1-9][0-9]*[MGT]i$",
              "description": "The monthly rate at which the volume grows."
            }
          },
          "required": ["startFrom", "monthlyRate"],
          "additionalProperties": false
        }
      },
      "required": ["type", "size"],
      "additionalProperties": false
    },
    "VolumeMount": {
      "type": "object",
      "properties": {
        "volume": {
          "type": "string",
          "pattern": "^(?!chainfile)[a-z0-9][a-z0-9-]{0,28}[a-z0-9]$",
          "description": "Name of the volume to mount, defined in the volumes section."
        },
        "mountPath": {
          "type": "string"
        },
        "subPath": {
          "type": "string"
        }
      },
      "required": ["volume", "mountPath"],
      "additionalProperties": false
    },
    "Container": {
      "type": "object",
      "properties": {
        "image": {
          "type": "string"
        },
        "tag": {
          "description": "Tag of the container image.",
          "oneOf": [
            {
              "type": "string"
            },
            {
              "$ref": "#/definitions/ParamReference"
            }
          ]
        },
        "source": {
          "type": "string",
          "description": "Source of the container image.",
          "format": "uri",
          "minLength": 1,
          "maxLength": 256
        },
        "description": {
          "type": "string",
          "minLength": 1,
          "maxLength": 1024
        },
        "endpoints": {
          "type": "object",
          "patternProperties": {
            "^(?!chainfile)[a-z0-9][a-z0-9-]{0,28}[a-z0-9]$": {
              "$ref": "#/definitions/Endpoint"
            }
          },
          "additionalProperties": false
        },
        "resources": {
          "type": "object",
          "description": "For placement of the node in a container cluster.",
          "properties": {
            "cpu": {
              "type": "number",
              "description": "CPU in vCPU (Virtual CPU Core, Relative). In multiples of 0.25, where 0.25 is 1/4 of a core.",
              "minimum": 0.25,
              "maximum": 16,
              "multipleOf": 0.25
            },
            "memory": {
              "type": "number",
              "description": "Memory in MiB (Mebibyte, Absolute). You are guaranteed to have this much memory available to your container.",
              "minimum": 64,
              "maximum": 122880,
              "multipleOf": 64
            }
          },
          "required": ["cpu", "memory"],
          "additionalProperties": false
        },
        "environment": {
          "type": "object",
          "maxProperties": 30,
          "patternProperties": {
            "^[a-zA-Z_]+[a-zA-Z0-9_]*$": {
              "oneOf": [
                {
                  "type": "string"
                },
                {
                  "$ref": "#/definitions/ParamReference"
                }
              ]
            }
          },
          "additionalProperties": false
        },
        "command": {
          "type": "array",
          "items": {
            "type": "string"
          }
        },
        "mounts": {
          "type": "array",
          "items": {
            "$ref": "#/definitions/VolumeMount"
          }
        }
      },
      "required": ["image", "tag", "source", "resources"],
      "additionalProperties": false
    },
    "Endpoint": {
      "type": "object",
      "description": "Extension to ports, endpoints are used to expose the container to the outside world with a specific protocol for interfacing.",
      "properties": {
        "description": {
          "type": "string",
          "minLength": 1,
          "maxLength": 1024
        }
      },
      "oneOf": [
        {
          "$ref": "#/definitions/EndpointPort"
        },
        {
          "$ref": "#/definitions/EndpointHttpRest"
        },
        {
          "$ref": "#/definitions/EndpointHttpJsonRpc"
        }
      ]
    },
    "EndpointPort": {
      "type": "object",
      "properties": {
        "port": {
          "type": "number",
          "minimum": 1,
          "maximum": 65535,
          "multipleOf": 1
        }
      },
      "required": ["port"],
      "additionalProperties": false
    },
    "EndpointHttpRest": {
      "type": "object",
      "properties": {
        "port": {
          "type": "number",
          "minimum": 1,
          "maximum": 65535,
          "multipleOf": 1
        },
        "protocol": {
          "enum": ["HTTP REST", "HTTPS REST"]
        },
        "authorization": {
          "$ref": "#/definitions/EndpointHttpAuthorization"
        },
        "probes": {
          "type": "object",
          "properties": {
            "readiness": {
              "description": "Probe to determine if the container is ready to receive traffic.",
              "$ref": "#/definitions/EndpointHttpRestProbe"
            },
            "liveness": {
              "description": "Probe to determine if the container is alive. Where an application is running, but unable to make progress this is useful to determine if the container should be restarted.",
              "$ref": "#/definitions/EndpointHttpRestProbe"
            },
            "startup": {
              "description": "Probe to determine if the container has started. This is useful for containers that take a long time to start up as it will prevent containers from being prematurely marked as unhealthy.",
              "$ref": "#/definitions/EndpointHttpRestProbe"
            }
          },
          "additionalProperties": false
        }
      },
      "required": ["port", "protocol"],
      "additionalProperties": false
    },
    "EndpointHttpRestProbe": {
      "type": "object",
      "properties": {
        "method": {
          "enum": ["GET", "POST", "PUT", "PATCH", "DELETE", "HEAD", "OPTIONS", "TRACE"]
        },
        "path": {
          "type": "string"
        },
        "body": {
          "type": "object"
        },
        "match": {
          "type": "object",
          "properties": {
            "status": {
              "oneOf": [
                {
                  "type": "number",
                  "minimum": 100,
                  "maximum": 599,
                  "multipleOf": 1
                },
                {
                  "type": "array",
                  "uniqueItems": true,
                  "items": {
                    "type": "number",
                    "minimum": 100,
                    "maximum": 599,
                    "multipleOf": 1
                  }
                }
              ]
            },
            "body": {
              "$ref": "http://json-schema.org/draft-07/schema#"
            }
          },
          "required": ["status"]
        }
      },
      "required": ["method", "path", "match"],
      "additionalProperties": false
    },
    "EndpointHttpJsonRpc": {
      "type": "object",
      "properties": {
        "port": {
          "type": "number",
          "minimum": 1,
          "maximum": 65535,
          "multipleOf": 1
        },
        "protocol": {
          "enum": ["HTTP JSON-RPC 1.0", "HTTPS JSON-RPC 1.0", "HTTP JSON-RPC 2.0", "HTTPS JSON-RPC 2.0"]
        },
        "path": {
          "type": "string"
        },
        "authorization": {
          "$ref": "#/definitions/EndpointHttpAuthorization"
        },
        "probes": {
          "type": "object",
          "properties": {
            "readiness": {
              "description": "Probe to determine if the container is ready to receive traffic.",
              "$ref": "#/definitions/EndpointHttpJsonRpcProbe"
            },
            "liveness": {
              "description": "Probe to determine if the container is alive. Where an application is running, but unable to make progress this is useful to determine if the container should be restarted.",
              "$ref": "#/definitions/EndpointHttpJsonRpcProbe"
            },
            "startup": {
              "description": "Probe to determine if the container has started. This is useful for containers that take a long time to start up as it will prevent containers from being prematurely marked as unhealthy.",
              "$ref": "#/definitions/EndpointHttpJsonRpcProbe"
            }
          },
          "additionalProperties": false
        }
      },
      "required": ["port", "protocol"],
      "additionalProperties": false
    },
    "EndpointHttpJsonRpcProbe": {
      "type": "object",
      "properties": {
        "method": {
          "type": "string"
        },
        "params": {
          "oneOf": [
            {
              "type": "array"
            },
            {
              "type": "object"
            }
          ]
        },
        "match": {
          "type": "object",
          "properties": {
            "result": {
              "$ref": "http://json-schema.org/draft-07/schema"
            }
          },
          "required": ["result"]
        }
      },
      "required": ["method", "params", "match"],
      "additionalProperties": false
    },
    "EndpointHttpAuthorization": {
      "oneOf": [
        {
          "type": "object",
          "properties": {
            "type": {
              "const": "HttpBasic"
            },
            "username": {
              "oneOf": [
                {
                  "type": "string"
                },
                {
                  "$ref": "#/definitions/ParamReference"
                }
              ]
            },
            "password": {
              "oneOf": [
                {
                  "type": "string"
                },
                {
                  "$ref": "#/definitions/ParamReference"
                }
              ]
            }
          },
          "required": ["type", "username", "password"],
          "additionalProperties": false
        },
        {
          "type": "object",
          "properties": {
            "type": {
              "const": "HttpBearer"
            },
            "token": {
              "oneOf": [
                {
                  "type": "string"
                },
                {
                  "$ref": "#/definitions/ParamReference"
                }
              ]
            }
          },
          "required": ["type", "token"],
          "additionalProperties": false
        }
      ]
    }
  }
}
