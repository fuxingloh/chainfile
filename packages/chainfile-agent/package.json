{
  "name": "@chainfile/agent",
  "version": "0.0.0",
  "private": false,
  "repository": {
    "url": "git+https://github.com/vetumorg/chainfile"
  },
  "license": "MIT",
  "scripts": {
    "build": "tsc --project tsconfig.build.json",
    "build:docker": "docker buildx build --progress=plain -t ghcr.io/vetumorg/chainfile-agent:$(node -p \"require('./package.json').version\") -f Dockerfile ../../",
    "clean": "rm -rf dist",
    "lint": "eslint .",
    "push:docker": "docker buildx build --progress=plain -t ghcr.io/vetumorg/chainfile-agent:$(node -p \"require('./package.json').version\") -f Dockerfile ../../ --output type=registry --platform linux/amd64,linux/arm64",
    "test": "jest"
  },
  "lint-staged": {
    "*": [
      "prettier --write --ignore-unknown"
    ],
    "*.{ts,tsx}": [
      "eslint --fix",
      "prettier --write"
    ]
  },
  "jest": {
    "preset": "@workspace/jest-preset"
  },
  "dependencies": {
    "@chainfile/schema": "workspace:^",
    "@trpc/server": "^10.45.2",
    "ajv": "^8.16.0",
    "ajv-formats": "^3.0.1",
    "debug": "^4.3.5",
    "trpc-openapi": "^1.2.0",
    "zod": "^3.23.8"
  },
  "devDependencies": {
    "@types/debug": "^4.1.12"
  }
}
