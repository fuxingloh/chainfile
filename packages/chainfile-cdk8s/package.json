{
  "name": "@chainfile/cdk8s",
  "version": "0.0.0",
  "private": false,
  "repository": {
    "url": "git+https://github.com/vetumorg/chainfile"
  },
  "license": "MIT",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist/*"
  ],
  "scripts": {
    "build": "tsc --project tsconfig.build.json",
    "clean": "rm -rf dist",
    "lint": "eslint .",
    "test": "jest"
  },
  "lint-staged": {
    "*": [
      "prettier --write --ignore-unknown"
    ],
    "*.{ts,tsx}": [
      "eslint --fix",
      "prettier --write"
    ]
  },
  "jest": {
    "preset": "@workspace/jest"
  },
  "dependencies": {
    "@chainfile/docker": "workspace:^",
    "@chainfile/schema": "workspace:^"
  },
  "devDependencies": {
    "cdk8s": "^2.68.84",
    "cdk8s-plus-25": "^2.22.79",
    "constructs": "^10.3.0",
    "get-port": "^5.1.1",
    "typescript": "4.9.5"
  },
  "peerDependencies": {
    "cdk8s": "^2.68.11",
    "cdk8s-plus-25": "^2.22.79",
    "constructs": "^10.3.0"
  }
}
