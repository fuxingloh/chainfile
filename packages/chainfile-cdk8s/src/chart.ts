import { Chainfile, validate } from '@chainfile/schema';
import { Chart } from 'cdk8s';
import { Construct } from 'constructs';

import { CFSecret } from './params';
import { CFService, CFServiceProps } from './service';
import { CFStatefulSet } from './sts';

export interface CFChartProps {
  chainfile: object;
  /**
   * Namespace to deploy the chart into.
   */
  namespace?: string;

  /**
   * Override params in Chainfile.
   * For parameters that use `default.random`
   * the value must be injected as CFChart does not support generating random values to ensure reproducibility.
   */
  params?: Record<string, string>;

  labels?: Record<string, string>;

  spec: {
    replicas?: number;
    storageClassName?: string;

    /**
     * Ports to expose on the service.
     * Each port must have a unique name.
     */
    exposes: CFServiceProps['spec']['ports'];
  };
}

export class CFChart extends Chart {
  private readonly service: CFService;

  constructor(scope: Construct, id: string, props: CFChartProps) {
    validate(props.chainfile);
    const chainfile = props.chainfile as Chainfile;
    const labels = props.labels ?? CFChart.getLabels(chainfile);

    super(scope, id, {
      namespace: props.namespace,
      labels: labels,
    });

    const secret = new CFSecret(this, 'secret', {
      chainfile: chainfile,
      params: props.params,
    });

    this.service = new CFService(this, 'service', {
      chainfile,
      metadata: {
        labels: labels,
      },
      spec: {
        type: 'LoadBalancer',
        selector: labels,
        ports: props.spec.exposes,
      },
    });

    new CFStatefulSet(this, 'stateful-set', {
      chainfile: chainfile,
      params: secret,
      metadata: {
        labels: labels,
      },
      spec: {
        template: {
          metadata: {
            labels: labels,
          },
        },
        // TODO(?): https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/
        serviceName: this.serviceName,
        replicas: props.spec.replicas,
        selector: {
          matchLabels: labels,
        },
        volumeClaimTemplates: {
          spec: {
            storageClassName: props.spec.storageClassName,
          },
        },
      },
    });
  }

  /**
   * Get the service name used in the chart.
   */
  public get serviceName(): string {
    return this.service.name;
  }

  /**
   * Get default labels from chainfile.
   * Which includes the name of chainfile, caip2, and available containers.
   *
   * More specificity can be implemented by using the container version, environment, etc.
   * But that is beyond the scope of a default selector.
   */
  public static getLabels(chainfile: Chainfile): Record<string, string> {
    // Regex: (([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?
    return {
      caip2: chainfile.caip2.replace(/[^A-Za-z0-9]/g, '.'),
      ...Object.fromEntries(Object.keys(chainfile.containers).map((name) => [name, 'true'])),
    };
  }
}
