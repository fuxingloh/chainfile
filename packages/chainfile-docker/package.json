{
  "name": "@chainfile/docker",
  "version": "0.0.0",
  "private": false,
  "repository": {
    "url": "git+https://github.com/vetumorg/chainfile"
  },
  "license": "MIT",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist/*"
  ],
  "scripts": {
    "build": "tsc --project tsconfig.build.json",
    "clean": "rm -rf dist",
    "lint": "eslint .",
    "test": "jest"
  },
  "lint-staged": {
    "*": [
      "prettier --write --ignore-unknown"
    ],
    "*.{ts,tsx}": [
      "eslint --fix",
      "prettier --write"
    ]
  },
  "jest": {
    "preset": "@workspace/jest"
  },
  "dependencies": {
    "@chainfile/schema": "workspace:^",
    "js-yaml": "^4.1.0"
  },
  "devDependencies": {
    "@types/js-yaml": "^4.0.9"
  }
}
