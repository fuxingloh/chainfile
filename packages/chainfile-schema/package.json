{
  "name": "@chainfile/schema",
  "version": "0.0.0",
  "private": false,
  "description": "A schema-oriented ecosystem packed with compute primitives such as cpu, memory, env, volume, snapshot, endpoint, and probes that can be composed to launch any blockchain node on any container-orchestration platform.",
  "repository": {
    "url": "git+https://github.com/fuxingloh/chainfile"
  },
  "license": "MIT",
  "main": "index.js",
  "types": "index.d.ts",
  "files": [
    "index.js",
    "index.d.ts",
    "schema.json",
    "schema.d.ts",
    "validate.js",
    "validate.d.ts"
  ],
  "scripts": {
    "build": "tsc --project tsconfig.build.json",
    "clean": "rm -rf schema.d.ts validate.d.ts validate.js",
    "generate": "json2ts -i schema.json -o schema.d.ts",
    "test": "jest"
  },
  "lint-staged": {
    "*": [
      "prettier --write --ignore-unknown"
    ],
    "*.{ts,tsx}": [
      "eslint --fix",
      "prettier --write"
    ]
  },
  "jest": {
    "preset": "@workspace/jest"
  },
  "dependencies": {
    "ajv": "^8.17.1",
    "ajv-formats": "^3.0.1"
  },
  "devDependencies": {
    "json-schema-to-typescript": "^14.1.0"
  }
}
