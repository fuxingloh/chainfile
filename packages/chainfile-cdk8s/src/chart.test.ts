import { execSync, spawn } from 'node:child_process';

import { Chainfile } from '@chainfile/schema';
import { afterAll, beforeAll, describe, expect, it, jest } from '@workspace/jest/globals';
import waitFor from '@workspace/jest/wait-for';
import { Testing } from 'cdk8s';
import getPort from 'get-port';

import { version } from '../package.json';
import { CFChart } from './chart';

global.Date.now = jest.fn(() => new Date('2024-01-01T00:00:00Z').getTime());

afterAll(() => {
  jest.restoreAllMocks();
});

const bitcoin_mainnet: Chainfile = {
  caip2: 'bip122:000000000019d6689c085ae165831e93',
  name: 'Bitcoin Mainnet',
  params: {
    rpc_user: {
      description: 'Username for RPC authentication',
      secret: true,
      default: {
        random: {
          bytes: 16,
          encoding: 'hex',
        },
      },
    },
    rpc_password: {
      description: 'Password for RPC authentication',
      secret: true,
      default: {
        random: {
          bytes: 16,
          encoding: 'hex',
        },
      },
    },
  },
  volumes: {
    data: {
      type: 'persistent',
      size: '600Gi',
      expansion: {
        startFrom: '2024-01-01',
        monthlyRate: '20Gi',
      },
    },
  },
  containers: {
    bitcoind: {
      image: 'docker.io/kylemanna/bitcoind',
      tag: 'latest',
      source: 'https://github.com/kylemanna/docker-bitcoind',
      endpoints: {
        p2p: {
          port: 8333,
        },
        rpc: {
          port: 8332,
          protocol: 'HTTP JSON-RPC 2.0',
          authorization: {
            type: 'HttpBasic',
            username: {
              $param: 'rpc_user',
            },
            password: {
              $param: 'rpc_password',
            },
          },
          probes: {
            readiness: {
              method: 'getblockchaininfo',
              params: [],
              match: {
                result: {
                  type: 'object',
                  properties: {
                    blocks: {
                      type: 'number',
                    },
                  },
                  required: ['blocks'],
                },
              },
            },
          },
        },
      },
      resources: {
        cpu: 1,
        memory: 2048,
      },
      environment: {
        DISABLEWALLET: '1',
        RPCUSER: {
          $param: 'rpc_user',
        },
        RPCPASSWORD: {
          $param: 'rpc_password',
        },
      },
      mounts: [
        {
          volume: 'data',
          mountPath: '/bitcoin/.bitcoin',
          subPath: 'bitcoind',
        },
      ],
    },
  },
};

describe('bitcoin_mainnet.json', () => {
  const app = Testing.app();
  const chart = new CFChart(app, 'bitcoin_mainnet.json', {
    chainfile: bitcoin_mainnet,
    params: {
      rpc_user: 'user',
      rpc_password: 'pass',
    },
    spec: {
      replicas: 2,
      exposes: [
        {
          name: 'http',
          port: 80,
          target: {
            container: 'bitcoind',
            endpoint: 'rpc',
          },
        },
      ],
    },
  });
  const results = Testing.synth(chart);

  it('should have 3 objects', async () => {
    expect(results).toHaveLength(3);
  });

  it('should synth bitcoin_mainnet.json and match snapshot', async () => {
    expect(results).toMatchSnapshot();
  });

  it('should get labels', async () => {
    expect(chart.labels).toStrictEqual({
      bitcoind: 'true',
      caip2: 'bip122.000000000019d6689c085ae165831e93',
    });
  });

  it('should get serviceName', async () => {
    expect(chart.serviceName).toMatch(/bitcoin_mainnet.json-service-[0-9a-f]{8}/);
  });
});

describe.skip('repl', () => {
  // Convenient tests to repeatedly apply chart to observe behavior
  // Set to skip by default
  const cluster = 'repl';
  const app = Testing.app();
  const chart = new CFChart(app, cluster, {
    chainfile: bitcoin_mainnet,
    params: {
      rpc_user: 'user',
      rpc_password: 'pass',
    },
    spec: {
      replicas: 1,
      exposes: [
        {
          port: 80,
          name: 'http',
          target: {
            container: 'bitcoind',
            endpoint: 'rpc',
          },
        },
      ],
    },
  });

  describe('cluster', () => {
    it('should create cluster', async () => {
      execSync(`kind create cluster --name ${cluster} --config kind.k8s.yaml`, { stdio: 'inherit' });
      execSync(`kind load docker-image docker.io/kylemanna/bitcoind:latest --name ${cluster}`, { stdio: 'inherit' });
      execSync(`kind load docker-image ghcr.io/vetumorg/chainfile-agent:${version} --name ${cluster}`, {
        stdio: 'inherit',
      });
    });

    it('should delete cluster', async () => {
      execSync(`kind delete cluster --name ${cluster}`, { stdio: 'inherit' });
    });
  });

  it('should apply chart', async () => {
    for (const resource of Testing.synth(chart)) {
      execSync(`kubectl apply --context kind-${cluster} -f -`, {
        input: JSON.stringify(resource),
        stdio: ['pipe', 'inherit', 'inherit'],
      });
    }
  });

  it('should delete chart', async () => {
    for (const resource of Testing.synth(chart)) {
      execSync(`kubectl delete --context kind-${cluster} -f -`, {
        input: JSON.stringify(resource),
        stdio: ['pipe', 'inherit', 'inherit'],
      });
    }
  });
});

describe('kind (k8s-in-docker)', () => {
  jest.setTimeout(600000);

  const cluster = 'cf-bitcoin-main';
  beforeAll(() => {
    execSync(`kind create cluster --name ${cluster} --config kind.k8s.yaml`, { stdio: 'inherit' });
    // Preload images from local environment to speed up tests
    execSync('docker pull kylemanna/bitcoind:latest', { stdio: 'inherit' });
    execSync(`kind load docker-image docker.io/kylemanna/bitcoind:latest --name ${cluster}`, { stdio: 'inherit' });
    execSync(`kind load docker-image ghcr.io/vetumorg/chainfile-agent:${version} --name ${cluster}`, {
      stdio: 'inherit',
    });
  });

  afterAll(() => {
    execSync(`kind delete cluster --name ${cluster}`, { stdio: 'inherit' });
  });

  it('should deploy bitcoin_mainnet.json chart and connect to service', async () => {
    const app = Testing.app();
    const chart = new CFChart(app, cluster, {
      chainfile: bitcoin_mainnet,
      params: {
        rpc_user: 'user',
        rpc_password: 'pass',
      },
      spec: {
        replicas: 1,
        exposes: [
          {
            port: 80,
            name: 'http',
            target: {
              container: 'bitcoind',
              endpoint: 'rpc',
            },
          },
        ],
      },
    });

    // Synth and apply the chart resources
    for (const resource of Testing.synth(chart)) {
      execSync(`kubectl apply --context kind-${cluster} -f -`, {
        input: JSON.stringify(resource),
        stdio: ['pipe', 'inherit', 'inherit'],
      });
    }

    process.stdout.write(
      `Waiting for bitcoind pod to be ready. You can also run 'kubectl describe pods --context kind-${cluster}' to check the status.\n`,
    );
    execSync(`kubectl wait --context kind-${cluster} --for=condition=Ready pod -l bitcoind=true --timeout=600s`, {
      stdio: 'inherit',
    });

    const port = await getPort();

    // This is a simple test to check if the pod is running.
    // Service routing can't actually be tested with kubectl port-forward.
    const forwarding = spawn('kubectl', ['port-forward', `service/${chart.serviceName}`, `${port}:http`], {
      stdio: 'inherit',
    });

    // Wait for port-forward to be ready, 405 because this endpoint only accepts POST
    await waitFor(async () => {
      const response = await fetch(`http://localhost:${port}/`);
      expect(response.status).toBe(405);
    }, 10000);

    await new Promise((resolve) => setTimeout(resolve, 1000));

    const response = await fetch(`http://localhost:${port}/`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: 'Basic ' + Buffer.from('user:pass').toString('base64'),
      },
      body: JSON.stringify({
        jsonrpc: '2.0',
        method: 'getblockchaininfo',
        params: [],
        id: 1,
      }),
    });

    expect(response.status).toBe(200);
    expect(await response.json()).toMatchObject({
      result: {
        chain: 'main',
      },
    });

    forwarding.kill('SIGTERM');
  });
});
