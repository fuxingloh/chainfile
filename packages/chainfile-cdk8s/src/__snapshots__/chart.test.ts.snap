// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bitcoin_mainnet.json should synth bitcoin_mainnet.json and match snapshot 1`] = `
[
  {
    "apiVersion": "v1",
    "kind": "Secret",
    "metadata": {
      "labels": {
        "bitcoind": "true",
        "caip2": "bip122.000000000019d6689c085ae165831e93",
      },
      "name": "bitcoin_mainnet.json-secret-c89840a1",
    },
    "stringData": {
      "CHAINFILE_PARAMS": "{"rpc_user":"user","rpc_password":"pass"}",
      "rpc_password": "pass",
      "rpc_user": "user",
    },
    "type": "Opaque",
  },
  {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": {
      "labels": {
        "bitcoind": "true",
        "caip2": "bip122.000000000019d6689c085ae165831e93",
      },
      "name": "bitcoin_mainnet.json-service-c8c0db95",
    },
    "spec": {
      "ports": [
        {
          "name": "http",
          "port": 80,
          "targetPort": 8332,
        },
      ],
      "selector": {
        "bitcoind": "true",
        "caip2": "bip122.000000000019d6689c085ae165831e93",
      },
      "type": "LoadBalancer",
    },
  },
  {
    "apiVersion": "apps/v1",
    "kind": "StatefulSet",
    "metadata": {
      "labels": {
        "bitcoind": "true",
        "caip2": "bip122.000000000019d6689c085ae165831e93",
      },
      "name": "bitcoin_mainnet.json-stateful-set-c8a4e78e",
    },
    "spec": {
      "replicas": 2,
      "selector": {
        "matchLabels": {
          "bitcoind": "true",
          "caip2": "bip122.000000000019d6689c085ae165831e93",
        },
      },
      "serviceName": "bitcoin_mainnet.json-service-c8c0db95",
      "template": {
        "metadata": {
          "labels": {
            "bitcoind": "true",
            "caip2": "bip122.000000000019d6689c085ae165831e93",
          },
        },
        "spec": {
          "containers": [
            {
              "env": [
                {
                  "name": "CHAINFILE_JSON",
                  "value": "{"caip2":"bip122:000000000019d6689c085ae165831e93","name":"Bitcoin Mainnet","params":{"rpc_user":{"description":"Username for RPC authentication","secret":true,"default":{"random":{"bytes":16,"encoding":"hex"}}},"rpc_password":{"description":"Password for RPC authentication","secret":true,"default":{"random":{"bytes":16,"encoding":"hex"}}}},"volumes":{"data":{"type":"persistent","size":"600Gi","expansion":{"startFrom":"2024-01-01","monthlyRate":"20Gi"}}},"containers":{"bitcoind":{"image":"docker.io/kylemanna/bitcoind","tag":"latest","source":"https://github.com/kylemanna/docker-bitcoind","endpoints":{"p2p":{"port":8333},"rpc":{"port":8332,"protocol":"HTTP JSON-RPC 2.0","authorization":{"type":"HttpBasic","username":{"$param":"rpc_user"},"password":{"$param":"rpc_password"}},"probes":{"readiness":{"method":"getblockchaininfo","params":[],"match":{"result":{"type":"object","properties":{"blocks":{"type":"number"}},"required":["blocks"]}}}}}},"resources":{"cpu":1,"memory":2048},"environment":{"DISABLEWALLET":"1","RPCUSER":{"$param":"rpc_user"},"RPCPASSWORD":{"$param":"rpc_password"}},"mounts":[{"volume":"data","mountPath":"/bitcoin/.bitcoin","subPath":"bitcoind"}]}}}",
                },
                {
                  "name": "CHAINFILE_PARAMS",
                  "valueFrom": {
                    "secretKeyRef": {
                      "key": "CHAINFILE_PARAMS",
                      "name": "bitcoin_mainnet.json-secret-c89840a1",
                      "optional": false,
                    },
                  },
                },
              ],
              "image": "ghcr.io/vetumorg/chainfile-agent:0.0.0",
              "livenessProbe": {
                "httpGet": {
                  "path": "/probes/liveness",
                  "port": "agent",
                },
              },
              "name": "agent",
              "ports": [
                {
                  "containerPort": 1569,
                  "name": "agent",
                },
              ],
              "readinessProbe": {
                "httpGet": {
                  "path": "/probes/readiness",
                  "port": "agent",
                },
              },
              "startupProbe": {
                "httpGet": {
                  "path": "/probes/startup",
                  "port": "agent",
                },
              },
            },
            {
              "env": [
                {
                  "name": "DISABLEWALLET",
                  "value": "1",
                },
                {
                  "name": "RPCUSER",
                  "valueFrom": {
                    "secretKeyRef": {
                      "key": "rpc_user",
                      "name": "bitcoin_mainnet.json-secret-c89840a1",
                      "optional": false,
                    },
                  },
                },
                {
                  "name": "RPCPASSWORD",
                  "valueFrom": {
                    "secretKeyRef": {
                      "key": "rpc_password",
                      "name": "bitcoin_mainnet.json-secret-c89840a1",
                      "optional": false,
                    },
                  },
                },
              ],
              "image": "docker.io/kylemanna/bitcoind:latest",
              "name": "bitcoind",
              "ports": [
                {
                  "containerPort": 8333,
                },
                {
                  "containerPort": 8332,
                },
              ],
              "resources": {
                "limits": {
                  "cpu": 1,
                  "memory": "2048Mi",
                },
              },
              "volumeMounts": [
                {
                  "mountPath": "/bitcoin/.bitcoin",
                  "name": "bitcoin_mainnet.json-pvc-data-c89840a1",
                  "subPath": "bitcoind",
                },
              ],
            },
          ],
          "hostAliases": [
            {
              "hostnames": [
                "bitcoind",
              ],
              "ip": "127.0.0.1",
            },
          ],
          "volumes": [],
        },
      },
      "volumeClaimTemplates": [
        {
          "metadata": {
            "name": "bitcoin_mainnet.json-pvc-data-c89840a1",
          },
          "spec": {
            "accessModes": [
              "ReadWriteOnce",
            ],
            "resources": {
              "requests": {
                "storage": "737280Mi",
              },
            },
          },
        },
      ],
    },
  },
]
`;
